content/courses/advanced-ada/chapters/types.rst

```diff
@@ -1212,16 +1212,434 @@ Data Representation
+seen above.
+
+Using Representation Clauses
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We can use representation clauses to change the way the compiler handles
+memory for a record type. For example, let's say we want to have an empty
+storage unit between components :ada:`A` and :ada:`B`. We can use a
+representation clause where we specify that component :ada:`B` starts at
+position #5 instead of #4, leaving an empty byte after component :ada:`A` and
+before component :ada:`B`:
+
+.. graphviz::
+
+    digraph foo {
+        "Record_R" [
+            label = "{ position | bits | component } |  { { { 0 | #0 .. 7 } | { 1 | #8 .. #15 } | { 2 | #16 .. #23 } | { 3 | #24 .. #31 } } | A } | { 4 |  |  } | { { { 5 | #0 .. 7 } | { 6 | #8 .. #15 } | { 7 | #16 .. #23 } | { 8 | #24 .. #31 } } | B }"
+            shape = "record"
+        ];
+   }
+
+This is the code that implements that:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Rep_Clauses_Empty_Byte
+
+    package P is
+
+       type R is record
+          A : Integer;
+          B : Integer;
+       end record;
+
+       for R use record
+          A at 0 range 0 .. 31;
+          B at 5 range 0 .. 31;
+       end record;
+
+    end P;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with P;           use P;
+
+    procedure Show_Empty_Byte is
+    begin
+       Put_Line ("R'Size:        " & R'Size'Image);
+       Put_Line ("R'Object_Size: " & R'Object_Size'Image);
+    end Show_Empty_Byte;
+
+When running the application above, we see that, due to the extra byte in the
+record representation, the sizes increase. On a typical 64-bit PC,
+:ada:`R'Size` is now 76 bits, which reflects the additional eight bits that we
+introduced between components :ada:`A` and :ada:`B`. Depending on the target
+architecture, you may also see that :ada:`R'Object_Size` is now 96 bits, which
+is the size the compiler selects as the most appropriate for this record type.
+As we've mentioned in the previous section, we can use aspects to request a
+specific size to the compiler. In this case, we could use the
+:ada:`Object_Size` aspect:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Rep_Clauses_Empty_Byte
+
+    package P is
+
+       type R is record
+          A : Integer;
+          B : Integer;
+       end record
+         with Object_Size => 72;
+
+       for R use record
+          A at 0 range 0 .. 31;
+          B at 5 range 0 .. 31;
+       end record;
+
+    end P;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with P;           use P;
+
+    procedure Show_Empty_Byte is
+    begin
+       Put_Line ("R'Size:        " & R'Size'Image);
+       Put_Line ("R'Object_Size: " & R'Object_Size'Image);
+    end Show_Empty_Byte;
+
+If the code compiles, :ada:`R'Size` and :ada:`R'Object_Size` should now have
+the same value.
+
+Derived Types And Representation Clauses
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In some cases, you might want to modify the memory representation of a record
```

Comment:
A common case for doing this is when you want to convert between the parent type and the derived type with a different representation. For example, one type might have a representation that is laid out for efficient access but therefore includes padding between some of the components.  We could have a compacted representation in the derived type, and use that to store the data on some external media. Using record representations and type conversions makes that easy.

Reply:
Yes, I agree, this is an interesting application. Actually, this topic is discussed in Robert Dewar's Gems 27 and 28, which are very good and therefore were included in this PR: 

https://github.com/AdaCore/learn/pull/575/files#diff-b2fee03b65a4b060b1eeede8faf1e0ba657e0b4971889376e1c54fb8546a134aR2028

That section (containing Robert's gems) is supposed to come after this one, so the goal here is to just provide a short introduction into representation clauses and leave the more interesting details to be discussed in Robert's article.